raw-spans-grouper/raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansGroupingTransformer.java

```diff
@@ -71,9 +95,29 @@ public void init(ProcessorContext context) {
   public KeyValue<String, StructuredTrace> transform(TraceIdentity key, RawSpan value) {
     ValueAndTimestamp<RawSpans> rawSpans = inflightTraceStore.get(key);
     RawSpans agg = rawSpans != null ? rawSpans.value() : RawSpans.newBuilder().build();
+    if (maxSpanCountMap.containsKey(key.getTenantId()) && agg.getRawSpans().size() > maxSpanCountMap.get(key.getTenantId())) {
+      if (logger.isDebugEnabled()) {
+        logger.debug("Number of spans in tenant_id={}, trace_id={} is {}",
+          key.getTenantId(),
+          HexUtils.getHex(key.getTraceId()),
+          agg.getRawSpans().size());
+        logger.debug("Dropping span [{}] from tenant_id={}, trace_id={}", value, key.getTenantId(), HexUtils.getHex(key.getTraceId()));
+      }
+      // increment the counter for dropped spans
+      droppedSpansCounter.computeIfAbsent(key.getTenantId(),
+        k -> PlatformMetricsRegistry.registerCounter(DROPPED_SPANS_COUNTER, Map.of("tenantId", k))).increment();
+      return null;
+    }
+
     // add the new span
     agg.getRawSpans().add(value);
 
+    // increment the counter when the trace size reaches the max.span.count limit.
+    if (maxSpanCountMap.containsKey(key.getTenantId()) && agg.getRawSpans().size() == maxSpanCountMap.get(key.getTenantId())) {
```

Comment:
why not track this along with the earlier condition ? For example if we have a limit of 100 and we have received 100 spans we track a truncated trace but we may not get the 101st span at all right ?

Reply:
In the previous if block, the condition will always be true from 101 spans onward leading to same trace being counted multiple times.